### PR TITLE
LPD-32067 language mixed in wc notification

### PR DESCRIPTION
**Pull request by marcogalluzzi**
# What is this trying to solve?

https://liferay.atlassian.net/browse/LPD-32067

The content of email notifications when Web Content is aded or updated, do not use the destination user language for macros such as [$ARTICLE_CONTENT$]

# How am I fixing it?

I'm using the method `setLocalizedContextAttribute` instead of `setContextAttribute` when adding the value of the macros to the `subscriptionSender` object.

# How can you verify that it works?

Follow the steps of the issue and/or execute the new tests created in class `JournalSubscriptionLocalizedContentTest`.

### CODE REVIEW

**Comment by marcogalluzzi** [comment]
The automatic tests are failing because of a text mismatch in the ASSERT due to some escaping that I'm not applying to the content generated from the test.

I don't know how to apply this escaping / sanitizing. Can you please help @jkappler ? The expelcted content is generated in `JournalSubscriptionLocalizedContentTest._getExpectedMailBody()`.

I know escaping can be done with `HtmlUtil.escape()` but this is not enough. See the differences below:

SITE_NAME = "Sitio de Pruebas en español" vs "Sitio de Pruebas en espa=C3=B1ol"

**Comment by jkappler** [comment]
@marcogalluzzi, taking a look

File: modules/apps/subscription/subscription-test-util/src/main/java/com/liferay/subscription/test/util/BaseSubscriptionLocalizedContentTestCase.java [hunk 1]
```diff
@@ -55,13 +54,13 @@ public void tearDown() throws Exception {
 	public void testSubscriptionLocalizedContentWhenAddingBaseModel()
 		throws Exception {
 
-		Map<Locale, String> previousLocalizedContents = HashMapBuilder.putAll(
```
Comment (jkappler) [inline modules/apps/subscription/subscription-test-util/src/main/java/com/liferay/subscription/test/util/BaseSubscriptionLocalizedContentTestCase.java:58]: @marcogalluzzi, isn't there a way to create the test without changing the base clases? They are used in too many places

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java [hunk 2]
```diff
@@ -44,14 +82,112 @@ public class JournalSubscriptionLocalizedContentTest
...
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), "Title"
+			).put(
+				LocaleUtil.SPAIN, "Título"
+			).build(),
+			null,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), "Content in English"
+			).put(
+				LocaleUtil.SPAIN, "Contenido en español"
```
Comment (jkappler) [inline modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:186]: @marcogalluzzi, there seems to be something off with special chars and the email server that we use, so I would suggest that we don't use special chars

**Review by jkappler — CHANGES_REQUESTED** [review]
Hi @marcogalluzzi, I'm unsure about the tests and changing the base classes just for this. 

Isn't there a way to avoid changing them? 

Also, it's a bit hard to follow the tests, mostly with how everything is set up and the class variables are used. Can we find a way to simplify this?

Thanks

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java [hunk 3]
```diff
@@ -44,14 +82,112 @@ public class JournalSubscriptionLocalizedContentTest
...
-			userId, group.getGroupId(), containerModelId);
+			group.getGroupId(), containerModelId,
+			JournalArticleConstants.CLASS_NAME_ID_DEFAULT,
+			HashMapBuilder.put(
+				LocaleUtil.getDefault(), "Title"
+			).put(
+				LocaleUtil.SPAIN, "Título"
+			).build(),
+			null,
+			HashMapBuilder.put(
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:183]: In the other tests GERMAN_BODY and SPANISH_BODY are used as a text for the entire body. In these tests I want to test how the macros are translated, so the new bodies includes macros.

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java [hunk 4]
```diff
@@ -113,10 +251,161 @@ PortletKeys.PREFS_PLID_SHARED, getServiceName(),
...
+		LayoutSet layoutSet = layout.getLayoutSet();
+
+		themeDisplay.setLayoutSet(layoutSet);
+
+		themeDisplay.setLayoutTypePortlet(
+			(LayoutTypePortlet)layout.getLayoutType());
+		themeDisplay.setLocale(LocaleUtil.getSiteDefault());
+		themeDisplay.setLookAndFeel(
+			layoutSet.getTheme(), layoutSet.getColorScheme());
+		themeDisplay.setRequest(_getHttpServletRequest(themeDisplay));
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:367]: It didn't fail for me, but I will set it. Thanks

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java [hunk 5]
```diff
@@ -44,14 +82,112 @@ public class JournalSubscriptionLocalizedContentTest
...
+		_themeDisplay = _getThemeDisplay(creatorUser);
+
+		long journalArticlePrimaryKey = addBaseModel(
+			creatorUser.getUserId(), getDefaultContainerModelId());
+
+		Assert.assertEquals(1, MailServiceTestUtil.getInboxSize());
+
+		MailMessage mailMessage = MailServiceTestUtil.getLastMailMessage();
+
+		_themeDisplay.setUser(user);
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:120]: Yes, I can do that. I wanted to save time and reuse it.

File: modules/apps/subscription/subscription-test-util/src/main/java/com/liferay/subscription/test/util/BaseSubscriptionLocalizedContentTestCase.java (hunk 1 above)
Comment (marcogalluzzi) [reply modules/apps/subscription/subscription-test-util/src/main/java/com/liferay/subscription/test/util/BaseSubscriptionLocalizedContentTestCase.java:58]: I've tested everywhere it is used. It shouldn't be any problem. The current implementation using the attribute `_localizedContent` seems to be unnecessary, this is why I'm proposing the change as an improvement.

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java (hunk 2 above)
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:186]: OK, I will avoid special characters. However, I don't know why HTML chars are escaped by the email server. This would avoid testing macro ARTICLE_DIFFS

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java]: Change done in [96a2798](https://github.com/liferay-content-management/liferay-portal/pull/5730/commits/96a2798c9ddf574307c3436c24b177f94ad6d1c7)

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java [hunk 6]
```diff
@@ -44,14 +82,112 @@ public class JournalSubscriptionLocalizedContentTest
...
+
 	@Override
 	protected long addBaseModel(long userId, long containerModelId)
 		throws Exception {
 
+		ServiceContext serviceContext =
+			ServiceContextTestUtil.getServiceContext(
+				group.getGroupId(), userId);
+
+		if (_themeDisplay != null) {
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java]: Change done in [96a2798](https://github.com/liferay-content-management/liferay-portal/pull/5730/commits/96a2798c9ddf574307c3436c24b177f94ad6d1c7)

File: modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java (hunk 2 above)
Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:201]: I've done this change in the test [dcc243b](https://github.com/liferay-content-management/liferay-portal/pull/5730/commits/dcc243bd23a77e215078332985cc4087025dc375). Please check if is OK because I'm not sure.

Comment (marcogalluzzi) [reply modules/apps/journal/journal-test/src/testIntegration/java/com/liferay/journal/subscriptions/test/JournalSubscriptionLocalizedContentTest.java:201]: Other solutions could be:

* Always set the `8bit` enconding in `MailEngine`. But I don't know if breaks some functionality.
* Set the "Content-Transfer-Encoding" header in the message when creating it in `JournalArticleLocalServiceImpl` by checking if the ServiceContext has it. This way I can pass this header from the test.

**Comment by marcogalluzzi** [comment]
ci:test:sf

File: modules/apps/journal/journal-service/src/main/java/com/liferay/journal/service/impl/JournalArticleLocalServiceImpl.java [hunk 7]
```diff
@@ -6816,52 +6817,27 @@ else if (action.equals("update") &&
...
+			new EscapableLocalizableFunction(
+				locale -> _getArticleContent(article, locale, serviceContext)));
+		subscriptionSender.setLocalizedContextAttribute(
+			"[$ARTICLE_STATUS$]",
+			new EscapableLocalizableFunction(
+				locale -> _language.get(
+					locale,
+					WorkflowConstants.getStatusLabel(article.getStatus()))));
+
+		JournalFolder folderCopy = folder;
```
Comment (jkappler) [inline modules/apps/journal/journal-service/src/main/java/com/liferay/journal/service/impl/JournalArticleLocalServiceImpl.java:6834]: @marcogalluzzi, do we need to create a copy of the folder? _getFolderName doesn't change it

**Comment by jkappler** [comment]
@marcogalluzzi, running relevant. Looks good, just the small question I asked

File: modules/apps/journal/journal-service/src/main/java/com/liferay/journal/service/impl/JournalArticleLocalServiceImpl.java (hunk 7 above)
Comment (marcogalluzzi) [reply modules/apps/journal/journal-service/src/main/java/com/liferay/journal/service/impl/JournalArticleLocalServiceImpl.java:6834]: @jkappler the copy is needed because the following lambda expression needs a constant or effectively constant variable. Is there some other pattern to use to avoid the copy?

**Comment by marcogalluzzi** [comment]
Unique fails in relevant are not related to my changes. Waiting for @jkappler to force forwarding.

**Review by jkappler — APPROVED** [review]

**Comment by jkappler** [comment]
ci:forward:force
